Governance update for Sandbox/Incubated/Graduated approach (#161)
camaraproject/Governance: ProjectStructureAndRoles.md

## Project Structure

**Overview**

The work of CAMARA is organized in 
* Sub Projects and their API Repositories
* Working Groups and their Repositories

API Repositories in CAMARA are classified with one of three stages of maturity:

* Sandboxed
* Incubated
* Graduated

Sandbox API Repositories can initially exist independent of a Sub Project. The process for the creation, incubation, graduation and (potential) archiving of API Repositories is described within [API-Onboarding-and-Lifecycle.md](./API-Onboarding-and-Lifecycle.md).

**Sub Projects**

> Sub Projects are the primary structure of the CAMARA Project. Each Sub Project is composed of Project participants from multiple companies and organizations, with a common purpose of advancing the Sub Project with respect to a specific Service API topic, for example ‘quality on demand’ or ‘localization’. Our goal is to enable distributed decision making and code ownership. This will be done by providing focused forums for getting work done, making decisions, and onboarding new Contributors.

Reviewer: Should we include here the term family? As it has evolved from its creation, right now families are mostly parallelizing the subgroup concept, a group of APIs with a relation among them but separated in different repositories to ensure that there is not cross-dependancy in the specification. Location, qod... are good examples for API families.

Reviewer 2: I'm a bit skeptical to re introduce the family term. We had discussion on the past but is was difficult to define and we get ride of it. Sub project term is just go enough for me.

Author: I'm here with Ludovic, we shouldn't use the term as an official term going forward as we have the structure elements of API Repository and Sub Projects with the "common purpose of advancing the Sub Project with respect to a specific Service API topic". We will get in addition the API Portfolio categories to structure the overall set of APIs within the communication towards API Consumers.

The wiki home page of a Sub Project should contain at least a description of the scope of the Sub Project, links to the API Repositories which are belonging to the Sub Project, the address of the mailing list on [lists.camaraproject.org](https://lists.camaraproject.org/groups), and the schedule and registation/join link of the regular meeting of the Sub Project. Sub pages should contain the meeting minutes, (informal) supporting documents and (if applicable) release information for the API members as defined by the Release Management working group.

A Sub Project should have at least three Maintainers. The Maintainers of a Sub Project are defined by the union of the Maintainers listed within the MAINTAINERS.md files of the API Repositories which are belonging to the Sub Project. It is recommended that there is an overlap between the Maintainers of the API Repositories of the Sub Project, but it is not mandated that all Maintainers of the Sub Project take responsibility for all APIs in the Sub Project.

> Each Sub Project is following for the API Repositories the Release Process as defined by the Release Management Working Group.

Reviewer 3: ```suggestion
Each Sub Project follows the Release Process for the API Repositories as defined by the Release Management Working Group.
```

**API Repositories**

The purpose of an API Repository is the definition and documentation of one or multiple (very closely related) APIs which will be always released together.

> An API Repository should have at least one Maintainer. Ideally an API Repository is managed by two or more Maintainers, depending on the size and scope of the API. Here, the responsibilities must be clearly agreed upon between all of the Maintainers. This includes coordinating who is responsible for which issues and pull requests. The Contributors to an API Repository are recorded by the mechanisms of GitHub in the Sub Project’s repositories.

Reviewer 3: For repositories which are not sandbox, is it enough to have 1 Maintainer?

Each API Repository should have subdirectories /documentation/API_documentation, /documentation/SupportingDocuments, /code/API_definitions, and /code/Test_definitions. 
  
API Repositories can have in addition Provider Implementation (PI) Repositories on their side. Provider Implementation Repositories only must have a CODEOWNERS file, a license file, a GOVERNANCE.MD file (pointing to the Governance repository) and a subdirectory /code/API_code with API code of the transformation function Northbound --> Southbound. Only Codeowners should have write permissions to a Provider Implementation (PI) Repository. Provider Implementation Repositories have the same name as the API Repository but with suffix _-PI§x§_.

**Sandbox API Repositories**

A Sandbox API Repository is meant for the rapid development of a new API proposal through one or multiple initial API versions (0.x.y) until the first public release of an initial or stable API Version within the context of a CAMARA Meta-release.

> 

Reviewer: Even when sandbox proposals are not detaily analized that will not occur until they have progressed and have a clearer scope and technical definition, We need to include basic rules that ensures that proposals included as sandbox, are still following the spirit of CAMARA.
```suggestion
Sandbox API Repositories 
```

Author: I agree with the intention of the suggested change, but would avoid to rephrase the scope (defined in the ProjectCharter) or technical API design guidelines in this document. Instead I added with 3b49da2449854875a7f8419f9fc3b58af6f1d1ce the following sentence:

> The work within Sandbox API Repositories - like in all API repositories - shall follow the governance and guidelines of CAMARA, including but not limited to the ProjectCharter, the Code of Conduct, the Contribution guidelines, the API Design Guidelines of Commonalities and the Release Process defined in Release Management.

Reviewer: Agreed, ok.

A Sandbox API Repository can be created independent of a CAMARA Sub Project or in context of an existing CAMARA Sub Project. It is the first stage for every new API proposal developed in CAMARA, for details see [API-Onboarding-and-Lifecycle.md](./API-Onboarding-and-Lifecycle.md).

> A Sandbox API Repository can belong to a CAMARA Sub Project or exist initially independent as an independent API Repository. 

Reviewer: Duplicated information with line 45

Author: Addressed in https://github.com/camaraproject/Governance/pull/161/commits/09911571145eb88a3dd3896a0151ac9f19f6b4c5

A Sandbox API Repository can be created independent of a CAMARA Sub Project or in context of an existing CAMARA Sub Project. It is the first stage for every new API proposal developed in CAMARA, for details see [API-Onboarding-and-Lifecycle.md](./API-Onboarding-and-Lifecycle.md).

A Sandbox API Repository can belong to a CAMARA Sub Project or exist initially independent. If the Sandbox API Repository belongs to a Sub Project it links within the README.md to the Sub Project wiki page. Mailing list, meetings will be for these Sandbox API Repositories organized in context of the Sub Project. A Sandbox API Repository which is not part of an existing Sub Project will get a Wiki page (in the Sandbox section), a preliminary mailing list, a Slack channel, and (on request) a Zoom meeting to be able to communicate with the community and attract contributors.

> A Sandbox API Repository must have at least one Codeowner and one Maintainer (could be the same person). If branch protection rules are activated (on request of the Maintainers) within the repository it must have at least two Codeowners. Branch protection rules must be get activated at latest when the first public release is created.

Reviewer 3: ```suggestion
A Sandbox API Repository must have at least one Codeowner and one Maintainer (could be the same person). If branch protection rules are activated (on request of the Maintainers) within the repository it must have at least two Codeowners. Branch protection rules must get activated at latest when the first public release is created.
```

Sandbox API Repositories are marked with a "Sandbox" badge in the README.md file and get a topic "Sandbox" assigned within the CAMARA GitHub Organization.

**Incubated API Repositories**

> An Incubated API Repository is meant for the further evolution of an API which has been implemented already by at least one operator, fits into the CAMARA API portfolio and gained sufficient support within the CAMARA community. For details of the Incubation process see [API-Onboarding-and-Lifecycle.md](./API-Onboarding-and-Lifecycle.md).

Reviewer 3: ```suggestion
An Incubated API Repository is meant for the further evolution of an API which has been implemented already by at least one operator, fits into the CAMARA API portfolio and has gained sufficient support within the CAMARA community. For details of the Incubation process see [API-Onboarding-and-Lifecycle.md](./API-Onboarding-and-Lifecycle.md).
```

> A Sandbox API Repository must have at least one Codeowner. If branch protection rules are activated within the repository it must have at least two Codeowners. Branch protection rules must get activated at latest when the first public release is created.

Reviewer: I'd delete the branch protection description here, since it's agreed that this not applies to sandbox repos.

Author: My idea here was that a public release should be protected, but I agree that this would restrict the work within a Sandbox. My view is that latest a stable API version should be maintained with branch protection, but that we can also define within Release Management. 

I changed the paragraph in https://github.com/camaraproject/Governance/pull/161/commits/d302e9501ae99f4d0d8ba729ef1e128282338c86 to:

> A Sandbox API Repository must have at least one Codeowner. If the Sandbox API Repository has two or more Codeowners the activation of branch protection rules is recommended.

An Incubated API Repository has the same structure as a Sandbox API Repository, with the following additional requirements: 
* An Incubated API Repository must belong to a CAMARA Sub Project (either an existing one or a new Sub Project created as part of the Incubation process). 

> * The repository should have at least three Maintainers (out of the group of Maintainers of the Sub Project it belongs to). 

Reviewer: is it necessary to specify if the maintainers need to belong to different companies here? we discussed about this in the draft, both for maintainers support and market adoption rules.

Reviewer 2: Agree with @jgarciahospital that this point should be added.

Author: Agree, addressed in f7082010fabe23a6a64817685672acdba3f5a05f

2 Either the issue creator or one of the Sub Project Maintainers then shall assign the issue to one of the Contributors of the Sub Project.

> 3 This Contributor shall create a fork (if not yet done before), shall create a new branch within the fork, and shall copy the branch link into the issue. For the naming of the branch a clear and concrete description shall be used.

Reviewer 3: Would copying the link in the issue be necessary? If we use the PR template and the #fixes tag, then this copying step is not needed. Probably not relevant for this PR but added this as this statement was part of the change.

Author: You are right, by using the "fixes" logic of GitHub, the PR will get automatically copied into the issue. But it is also possible to Lin the branch already earlier (there is a field within a GitHub issue for that, which is e.g. filled if the branch will be created out of the issue). Making this optional could be the scope of a separate PR.

**API Backlog**

The API Backlog Working Group is supervised by the TSC. The responsibilities of the Working Group are to take the decisions about API Proposals, including the creation of (Sandbox) API Repositories, to support the Incubation and Graduation process of API Repositories, to prepare the material for the creation of new Sub Projects or scope changes of existing Sub Project (e.g. adding new API Repositories to a Sub Project).

The decision about new Sandbox API Repositories will be taken by the Working Group. Decision about Incubation, Graduation, and onboarding of new Sub Projects will be taken by the TSC. Changes of existing Sub Projects have to be approved by the involved Sub Projects.

> The API onboarding and API lifecycle process in CAMARA will described in [API-Onboarding-and-Lifecycle.md](./API-Onboarding-and-Lifecycle.md).

Reviewer 3: ```suggestion
The API onboarding and API lifecycle process in CAMARA are described in [API-Onboarding-and-Lifecycle.md](./API-Onboarding-and-Lifecycle.md).
```
